internals/secrethub/audit.go

```diff
@@ -82,30 +129,170 @@ func (cmd *AuditCommand) run() error {
+// with the configured field names as keys and the provided values
+func (f *jsonFormatter) WriteRow(row []string) error {
+	if len(f.fields) != len(row) {
+		return fmt.Errorf("unexpected number of json fields")
+	}
+
+	jsonMap := make(map[string]string)
+	for i, element := range row {
+		jsonMap[f.fields[i]] = element
+	}
+
+	return f.encoder.Encode(jsonMap)
+}
+
+// newColumnFormatter returns a table formatter that aligns the columns of the table.
+func newColumnFormatter(writer io.Writer, tableWidth int, columns []tableColumn) *columnFormatter {
+	return &columnFormatter{writer: writer, tableWidth: tableWidth, columns: columns}
+}
+
+type columnFormatter struct {
+	tableWidth           int
+	writer               io.Writer
+	computedColumnWidths []int
+	columns              []tableColumn
+	didPrintHeader       bool
+}
+
+func (f *columnFormatter) WriteRow(row []string) error {
+	if !f.didPrintHeader {
+		header := make([]string, len(f.columns))
+		for i, col := range f.columns {
+			header[i] = col.name
 		}
+		formattedHeader := f.formatRow(header)
+		_, err := f.writer.Write(formattedHeader)
+		if err != nil {
+			return err
+		}
+		f.didPrintHeader = true
 	}
+	formattedRow := f.formatRow(row)
+	_, err := f.writer.Write(formattedRow)
+	return err
+}
 
-	err = tabWriter.Flush()
-	if err != nil {
-		return err
+// formatRow formats the given table row to fit the configured width by
+// giving each cell an equal width and wrapping the text in cells that exceed it.
+func (f *columnFormatter) formatRow(row []string) []byte {
+	columnWidths := f.columnWidths()
+
+	// calculate the maximum number of lines a cell value will be broken into
+	maxLinesPerCell := 1
+	for i, cell := range row {
+		lines := len(cell) / columnWidths[i]
+		if len(cell)%columnWidths[i] != 0 {
+			lines++
+		}
+		if lines > maxLinesPerCell {
+			maxLinesPerCell = lines
+		}
 	}
 
-	return nil
+	// split the cell values into a grid according to how they will be printed
+	splitCells := make([][]string, maxLinesPerCell)
+	for i := 0; i < maxLinesPerCell; i++ {
+		splitCells[i] = make([]string, len(row))
+	}
+
+	for i, cell := range row {
+		columnWidth := columnWidths[i]
+		lineCount := len(cell) / columnWidth
+		for j := 0; j < lineCount; j++ {
+			begin := j * columnWidth
+			end := (j + 1) * columnWidth
+			splitCells[j][i] = cell[begin:end]
+		}
+
+		charactersLeft := len(cell) % columnWidth
+		if charactersLeft != 0 {
+			splitCells[lineCount][i] = cell[len(cell)-charactersLeft:] + strings.Repeat(" ", columnWidth-charactersLeft)
+		} else if lineCount < maxLinesPerCell {
+			splitCells[lineCount][i] = strings.Repeat(" ", columnWidth)
+		}
+
+		for j := lineCount + 1; j < maxLinesPerCell; j++ {
+			splitCells[j][i] = strings.Repeat(" ", columnWidth)
+		}
+	}
+
+	// convert the grid to a string
+	strRes := strings.Builder{}
+	for j := 0; j < maxLinesPerCell; j++ {
+		strRes.WriteString(strings.Join(splitCells[j], "  ") + "\n")
+	}
+	return []byte(strRes.String())
```

Comment:
This is a really long function. Splitting it up into multiple functions has some benefits:

1. It forces you to limit the contact surface between different sections.
2. It self-describes the functionality.
3. Allows for easier testing (not always necessary)

In this case, I think you already spotted where the functions should be: at the places where you placed comments.

I haven't checked whether it is actually possible in this case, but give it a try.

Reply:
It should be quite straightforward. Thanks for explaining the motivation behind the requested change. This really helps me avoid repeating mistakes / bad habits such as this one.

Reply:
However I think splitting up this function might also have a disadvantage:
Most of the smaller functions don't make much sense on their own (e.g. `rowToGrid`, or `gridToString`) and can even be distracting. They are also heavily tied to the solution I came up with and are not fundamental to the problem of formatting a table. I think this would make more sense if the formatter was in a different package and these helper functions were not exported, because seeing them as methods on a formatter makes me wonder why they exist.

That being said I understand the advantage of reducing contact surface, making testing easier and increasing readability, so I'll implement the change. But I'm curious about your opinion on this matter, @jpcoenen

Reply:
I understand your concerns.

> They are also heavily tied to the solution I came up with and are not fundamental to the problem of formatting a table.

That's very true. The question is whether that is a problem. The methods you created are unexported, so in my opinion it's not a problem that they are specific to the underlying implementation. If there were exported, it would be another story and it would indeed be really smelly.

As you already note yourself: Go does not have a very tight mechanism for keeping functionality private to objects except for moving it to another package. I regularly run into situations where I'd like something to be private from other functions within the package. However, that's something that Go is not really built for. The main focus of the exported/unexported mechanism seems to be on hiding internal logic to _external_ users.

> I think this would make more sense if the formatter was in a different package and these helper functions were not exported, because seeing them as methods on a formatter makes me wonder why they exist.

That's not a bad idea. If there's a clear cut you can make, this is definitely an option. However, something to always keep in mind is not to create new packages just because you want a few functions/fields to be unaccessible; that quickly leads to the madness of having _many_ very small packages. Nevertheless, in this case it does appear to be contained and reusable logic, so a new package does seem appropriate.

> Most of the smaller functions don't make much sense on their own (e.g. rowToGrid) and can even be distracting. 

I agree on the `rowToGrid` part. Often that's a sign that:

1. The function does not have a clear, single responsibility, hence it cannot be described in a name. Fix: split or restructure the function.
1. The name does not really describe what is going on. What could work in this case is trying to explain to yourself what the code does. If you find yourself saying "and" a lot while doing this, refer to 1.

In this specific case, I think it's the latter. 

One thing I can imagine helps in naming here, is to rename the variable `row` to `cells` (at least if I correctly understand the purpose of this cell). Then the purpose of the function is easier to describe: "fit the content of cells to the given column widths" (again: correct me if I am wrong here). Then the function could be named something like: `fitCellsToColumns` or maybe even `fitToColumns` (because `cells` is implied in the argument). Does that help?

Reply:
Thanks, this is great feedback. The trade offs are quite clear now. I'll move the pagination logic to a different package, as suggested.

Reply:
Yes, it's a lot more readable now :grin:

Reply:
Did you have a look at my point on how to possibly improve the naming of/within `writeToGrid`?